gguf-py/gguf/constants.py

```diff
@@ -936,6 +935,366 @@ def get_type(val: Any) -> GGUFValueType:
+        "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}{1,3}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
+    ),
+    "falcon": (
+        "[\\p{P}\\$\\+<=>\\^~\\|]+", BPE_PRE_TOKENIZER_DEFAULT, "[0-9][0-9][0-9]",
+    ),
+    "mpt": (
+        "\\s?\\p{L}+", "\\s?\\p{P}+", BPE_PRE_TOKENIZER_DEFAULT,
+    ),
+    # starcoder, refact, command-r
+    "starcoder": (
+        "\\p{N}", BPE_PRE_TOKENIZER_DEFAULT,
+    ),
+    # qwen, qwen2, stablelm
+    "qwen": (
+        "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
+    ),
+    # NOTE: deepseek uses the 'llama' arch, but diverges with the pre-tok.
+    "deepseek": (
+        "[\r\n]",
+        "\\s?[A-Za-zµÀ-ÖØ-öø-ƺƼ-ƿǄ-ʓʕ-ʯͰ-ͳͶͷͻ-ͽͿΆΈ-ΊΌΎ-ΡΣ-ϵϷ-ҁҊ-ԯԱ-ՖႠ-ჅᎠ-Ᏽᏸ-ᏽᲐ-ᲺᲽ-Ჿᴀ-ᴫᵫ-ᵷᵹ-ᶚḀ-ἕἘ-Ἕἠ-ὅὈ-Ὅὐ-ὗὙὛὝὟ-ώᾀ-ᾴᾶ-ᾼιῂ-ῄῆ-ῌῐ-ΐῖ-Ίῠ-Ῥῲ-ῴῶ-ῼℂℇℊ-ℓℕℙ-ℝℤΩℨK-ℭℯ-ℴℹℼ-ℿⅅ-ⅉⅎↃↄⰀ-ⱻⱾ-ⳤⳫ-ⳮⳲⳳꙀ-ꙭꚀ-ꚛꜢ-ꝯꝱ-ꞇꞋ-ꞎꭰ-ꮿﬀ-ﬆﬓ-ﬗＡ-Ｚａ-ｚ𐐀-𐑏𐒰-𐓓𐓘-𐓻𐲀-𐲲𐳀-𐳲𑢠-𑣟𞤀-𞥃]+",
+        "\\s?[!-/:-~！-／：-～‘-‟　-。]+",
+        "\\s+$",
+        "[一-龥ࠀ-一가-퟿]+",
+        "\\p{N}+",
+    ),
+    "deepseek-coder": (
+        "[\r\n]",
+        "\\s?\\p{L}+",
+        "\\s?\\p{P}+",
+        "[一-龥ࠀ-一가-퟿]+",
+        "\\p{N}",
+    ),
+    # NOTE: ONLY ADD MODELS ON A AS NEEDED BASIS.
+    # This will get out of control if not properly managed.
+    # This needs a proper solution. The short-term solution is to manually build a map here.
+    # A proper long-term solution would be to build a dynamic registry.
+    # The issue is that this requires a dynamically persistent mapping or a database.
+    # Possible solutions are to use JSON, HDF5, or SQLite.
+    # Some of these mappings could be dynamically generated, but it's sketchy at best.
+    # Model versions should be included along with the model name to mitigate name conflicts.
+    # This entire setup is extremely fragile and will result in breaking changes in the future.
+}
+
+#
+# HuggingFace Model Map
+#
+# NOTE: All prerequisite model metadata must be defined here.
+#
+# Defines metadata for each Hugging Face model required during conversion to GGUF
+#
+# Field Descriptions
+#   - `model_repo` (str): The HuggingFace endpoint or local path to the models repository
+#   - `model_arch` (MODEL_ARCH): Model architecture type
+#   - `model_parts` (int): Number of parts required to join the model during conversion
+#   - `model_type` (FileFormatType): File format for the Hugging Face model files
+#   - `vocab_type` (VocabType): Vocabulary type used by the tokenizer
+#   - `vocab_pre` (Optional[Tuple[str]]): Tuple of pre-tokenizer pattern strings for this model
+#   - `vocab_files` (Tuple[str]): Tuple of file names required to extract vocabulary and other metadata
+#
+# NOTES
+#   - Possible algorithms are WordLevel, BPE, WordPiece, or Unigram
+#   - Possible LLaMa tokenizer model types are: None, SPM, BPE, or WPM
+HF_MODEL_MAP = (
```

Comment:
@mofosyne I'm thinking about adding a registry. This way we can register the necessary metadata for each model. The easiest way I can think about doing it for now is to write it to a JSON file.